activesupport/lib/active_support/callbacks.rb

```diff
@@ -826,6 +754,30 @@ def get_callbacks(name)
       def set_callbacks(name, callbacks)
         send "_#{name}_callbacks=", callbacks
       end
+
+      def default_false_terminator
+        Proc.new do |target, result_lambda|
+          terminate = true
+          catch(:abort) do
+            result = result_lambda.call if result_lambda.is_a?(Proc)
+            if CallbackChain.halt_and_display_warning_on_return_false && result == false
+              display_deprecation_warning_for_false_terminator
+            else
+              terminate = false
+            end
+          end
+          terminate
+        end
+      end
+
+      private
+
+      def display_deprecation_warning_for_false_terminator
+        ActiveSupport::Deprecation.warn(<<-MSG.squish)
+          Returning `false` in a callback will not implicitly halt a callback chain in the next release of Rails.
+          To explicitly halt a callback chain, please use `throw :abort` instead.
```

Comment:
And finally… this message would be much clearer like this:

```ruby
ActiveSupport::Deprecation.warn(<<-MSG.squish)
  Returning `false` in an #{klass} callback will not implicitly halt a callback chain in the next release of Rails.
  To explicitly halt an #{klass} callback chain, please use `throw :abort` instead.
MSG
```

So the users would understand that the behavior has changed **only** for ActiveModel, ActiveModel::Validations and ActiveRecord, and not for any other third-party app using ActiveSupport::Callbacks